File: generate-meal-plan-pdf.js
import fs from 'fs';
import PDFDocument from 'pdfkit';

function parseMealPlan(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    // Handle both \n and \r\n line endings
    const lines = content.split(/\r?\n/);
    
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    const mealPlan = {};
    let currentDay = null;
    
    // Initialize all days
    days.forEach(day => {
        mealPlan[day] = {
            'Breakfast': '',
            'Mid-Morning Snack': '',
            'Lunch': '',
            'Afternoon Snack': '',
            'Dinner': '',
            'Evening': ''
        };
    });
    
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const trimmed = line.trim();
        if (!trimmed) continue;
        
        // Skip the title line
        if (trimmed.toUpperCase().includes('SAMPLE') || trimmed.toUpperCase().includes('MEAL PLAN')) {
            continue;
        }
        
        // Check if it's a day name (exact match)
        if (days.includes(trimmed)) {
            currentDay = trimmed;
            continue;
        }
        
        // Process meal entries
        if (currentDay && mealPlan[currentDay]) {
            if (trimmed.startsWith('Breakfast:')) {
                mealPlan[currentDay]['Breakfast'] = trimmed.replace(/^Breakfast:\s*/, '').trim();
            } else if (trimmed.startsWith('Mid-Morning Snack:')) {
                mealPlan[currentDay]['Mid-Morning Snack'] = trimmed.replace(/^Mid-Morning Snack:\s*/, '').trim();
            } else if (trimmed.startsWith('Lunch:')) {
                mealPlan[currentDay]['Lunch'] = trimmed.replace(/^Lunch:\s*/, '').trim();
            } else if (trimmed.startsWith('Afternoon Snack:')) {
                mealPlan[currentDay]['Afternoon Snack'] = trimmed.replace(/^Afternoon Snack:\s*/, '').trim();
            } else if (trimmed.startsWith('Dinner:')) {
                mealPlan[currentDay]['Dinner'] = trimmed.replace(/^Dinner:\s*/, '').trim();
            } else if (trimmed.startsWith('Evening')) {
                // Handle both "Evening:" and "Evening (if needed):"
                mealPlan[currentDay]['Evening'] = trimmed.replace(/^Evening\s*\([^)]*\)?\s*:\s*/, '').trim();
            }
        }
    }
    
    // Debug: Log parsed data
    console.log('Parsed meal plan sample:');
    console.log('Monday Breakfast:', mealPlan['Monday']['Breakfast'] || '(empty)');
    console.log('Monday Lunch:', mealPlan['Monday']['Lunch'] || '(empty)');
    
    return mealPlan;
}

function wrapText(text, maxWidth, doc, fontSize) {
    const words = text.split(' ');
    const lines = [];
    let currentLine = '';
    
    for (const word of words) {
        const testLine = currentLine ? `${currentLine} ${word}` : word;
        const width = doc.widthOfString(testLine, { fontSize });
        
        if (width > maxWidth && currentLine) {
            lines.push(currentLine);
            currentLine = word;
        } else {
            currentLine = testLine;
        }
    }
    
    if (currentLine) {
        lines.push(currentLine);
    }
    
    return lines.length > 0 ? lines : [''];
}

function getMaxLines(rowData, colWidths, doc, fontSize) {
    let maxLines = 1;
    
    rowData.forEach((cell, idx) => {
        const wrapped = wrapText(cell || '', colWidths[idx] - 6, doc, fontSize);
        maxLines = Math.max(maxLines, wrapped.length);
    });
    
    return maxLines;
}

function createPDF(mealPlan, outputPath) {
    const doc = new PDFDocument({
        size: 'LETTER',
        margins: { top: 40, bottom: 40, left: 30, right: 30 }
    });
    
    doc.pipe(fs.createWriteStream(outputPath));
    
    // Title
    doc.fontSize(16)
       .font('Helvetica-Bold')
       .fillColor('#1a1a1a')
       .text('7-DAY MEAL PLAN WITH FRUITS', {
           align: 'center'
       });
    
    // Table setup
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    const columns = ['Day', 'Breakfast', 'Mid-Morning Snack', 'Lunch', 'Afternoon Snack', 'Dinner', 'Evening'];
    
    // Calculate column widths - optimized for single page
    const pageWidth = 612; // Letter size width in points
    const margins = 60; // Left + right margins
    const availableWidth = pageWidth - margins;
    const colWidths = [
        60,  // Day
        95,  // Breakfast
        80,  // Mid-Morning Snack
        95,  // Lunch
        80,  // Afternoon Snack
        95,  // Dinner
        55   // Evening
    ];
    
    const startX = 30;
    let startY = 70;
    const baseRowHeight = 20;
    const headerHeight = 30;
    const lineHeight = 8;
    const cellPadding = 3;
    const fontSize = 6.5;
    
    // Draw header
    doc.fontSize(8)
       .font('Helvetica-Bold')
       .fillColor('#FFFFFF');
    
    let currentX = startX;
    columns.forEach((col, idx) => {
        doc.rect(currentX, startY, colWidths[idx], headerHeight)
           .fill('#2c3e50');
        
        // Split long column names
        let colName = col;
        if (col === 'Mid-Morning Snack') {
            colName = 'Mid-\nMorning\nSnack';
        } else if (col === 'Afternoon Snack') {
            colName = 'After-\nnoon\nSnack';
        }
        const lines = colName.split('\n');
        
        let textY = startY + (headerHeight - (lines.length * 8)) / 2;
        lines.forEach((line, lineIdx) => {
            doc.fillColor('#FFFFFF')
               .text(line, currentX + cellPadding, textY + (lineIdx * 8), {
                   width: colWidths[idx] - (cellPadding * 2),
                   align: 'center'
               });
        });
        
        currentX += colWidths[idx];
    });
    
    startY += headerHeight;
    
    // Draw data rows
    doc.fontSize(fontSize)
       .font('Helvetica');
    
    days.forEach((day, rowIdx) => {
        const isEven = rowIdx % 2 === 0;
        const bgColor = isEven ? '#FFFFFF' : '#f8f9fa';
        
        const rowData = [
            day,
            mealPlan[day]['Breakfast'],
            mealPlan[day]['Mid-Morning Snack'],
            mealPlan[day]['Lunch'],
            mealPlan[day]['Afternoon Snack'],
            mealPlan[day]['Dinner'],
            mealPlan[day]['Evening']
        ];
        
        // Calculate row height based on content
        const maxLines = getMaxLines(rowData, colWidths, doc, fontSize);
        const rowHeight = Math.max(baseRowHeight, (maxLines * lineHeight) + (cellPadding * 2));
        
        // Ensure we don't go past page height
        const pageHeight = 792; // Letter size height
        if (startY + rowHeight > pageHeight - 40) {
            // Adjust row height to fit
            const remainingSpace = pageHeight - startY - 40;
            const adjustedRowHeight = Math.min(rowHeight, remainingSpace);
            const adjustedMaxLines = Math.floor((adjustedRowHeight - (cellPadding * 2)) / lineHeight);
            
            // Draw row background
            doc.rect(startX, startY, availableWidth, adjustedRowHeight)
               .fill(bgColor);
            
            // Draw borders
            doc.strokeColor('#CCCCCC')
               .lineWidth(0.5)
               .rect(startX, startY, availableWidth, adjustedRowHeight)
               .stroke();
            
            // Draw cell borders
            currentX = startX;
            colWidths.forEach((width, idx) => {
                if (idx < colWidths.length - 1) {
                    doc.moveTo(currentX + width, startY)
                       .lineTo(currentX + width, startY + adjustedRowHeight)
                       .stroke();
                }
                currentX += width;
            });
            
            // Draw cell content with truncation
            currentX = startX;
            rowData.forEach((cell, idx) => {
                const wrapped = wrapText(cell || '', colWidths[idx] - (cellPadding * 2), doc, fontSize);
                const displayLines = wrapped.slice(0, adjustedMaxLines);
                
                displayLines.forEach((line, lineIdx) => {
                    doc.fillColor('#000000')
                       .text(line, currentX + cellPadding, startY + cellPadding + (lineIdx * lineHeight), {
                           width: colWidths[idx] - (cellPadding * 2),
                           align: 'left'
                       });
                });
                
                currentX += colWidths[idx];
            });
            
            startY += adjustedRowHeight;
        } else {
            // Draw row background
            doc.rect(startX, startY, availableWidth, rowHeight)
               .fill(bgColor);
            
            // Draw borders
            doc.strokeColor('#CCCCCC')
               .lineWidth(0.5)
               .rect(startX, startY, availableWidth, rowHeight)
               .stroke();
            
            // Draw cell borders
            currentX = startX;
            colWidths.forEach((width, idx) => {
                if (idx < colWidths.length - 1) {
                    doc.moveTo(currentX + width, startY)
                       .lineTo(currentX + width, startY + rowHeight)
                       .stroke();
                }
                currentX += width;
            });
            
            // Draw cell content
            currentX = startX;
            rowData.forEach((cell, idx) => {
                const wrapped = wrapText(cell || '', colWidths[idx] - (cellPadding * 2), doc, fontSize);
                
                wrapped.forEach((line, lineIdx) => {
                    if (lineIdx * lineHeight < rowHeight - (cellPadding * 2)) {
                        doc.fillColor('#000000')
                           .text(line, currentX + cellPadding, startY + cellPadding + (lineIdx * lineHeight), {
                               width: colWidths[idx] - (cellPadding * 2),
                               align: 'left'
                           });
                    }
                });
                
                currentX += colWidths[idx];
            });
            
            startY += rowHeight;
        }
    });
    
    doc.end();
}

// Main execution
const inputFile = 'meal_plan.html';
const outputFile = 'meal_plan.pdf';

console.log(`Parsing meal plan from ${inputFile}...`);
const mealPlan = parseMealPlan(inputFile);

console.log(`Generating PDF: ${outputFile}...`);
createPDF(mealPlan, outputFile);

console.log(`✓ PDF created successfully: ${outputFile}`);

File: generate_meal_plan_pdf.py
#!/usr/bin/env python3
"""
Script to convert meal plan from text to a tabularized PDF
"""

from reportlab.lib import colors
from reportlab.lib.pagesizes import letter, A4
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.units import inch
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
import re

def parse_meal_plan(file_path):
    """Parse the meal plan text file into structured data"""
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
    meal_plan = {}
    current_day = None
    
    lines = content.split('\n')
    
    for line in lines:
        line = line.strip()
        if not line:
            continue
        
        # Check if it's a day
        if line in days:
            current_day = line
            meal_plan[current_day] = {
                'Breakfast': '',
                'Mid-Morning Snack': '',
                'Lunch': '',
                'Afternoon Snack': '',
                'Dinner': '',
                'Evening': ''
            }
        elif current_day:
            # Check for meal types
            if line.startswith('Breakfast:'):
                meal_plan[current_day]['Breakfast'] = line.replace('Breakfast:', '').strip()
            elif line.startswith('Mid-Morning Snack:'):
                meal_plan[current_day]['Mid-Morning Snack'] = line.replace('Mid-Morning Snack:', '').strip()
            elif line.startswith('Lunch:'):
                meal_plan[current_day]['Lunch'] = line.replace('Lunch:', '').strip()
            elif line.startswith('Afternoon Snack:'):
                meal_plan[current_day]['Afternoon Snack'] = line.replace('Afternoon Snack:', '').strip()
            elif line.startswith('Dinner:'):
                meal_plan[current_day]['Dinner'] = line.replace('Dinner:', '').strip()
            elif line.startswith('Evening') or line.startswith('Evening (if needed):'):
                meal_plan[current_day]['Evening'] = re.sub(r'^Evening.*?:', '', line).strip()
    
    return meal_plan

def create_pdf(meal_plan, output_path):
    """Create a PDF with the meal plan in table format"""
    doc = SimpleDocTemplate(output_path, pagesize=letter,
                           rightMargin=0.5*inch, leftMargin=0.5*inch,
                           topMargin=0.5*inch, bottomMargin=0.5*inch)
    
    # Container for the 'Flowable' objects
    elements = []
    
    # Title
    title_style = ParagraphStyle(
        'CustomTitle',
        parent=getSampleStyleSheet()['Heading1'],
        fontSize=18,
        textColor=colors.HexColor('#1a1a1a'),
        spaceAfter=20,
        alignment=1  # Center alignment
    )
    title = Paragraph("7-DAY MEAL PLAN WITH FRUITS", title_style)
    elements.append(title)
    elements.append(Spacer(1, 0.2*inch))
    
    # Prepare table data
    table_data = []
    
    # Header row
    header = ['Day', 'Breakfast', 'Mid-Morning<br/>Snack', 'Lunch', 
              'Afternoon<br/>Snack', 'Dinner', 'Evening']
    table_data.append(header)
    
    # Data rows
    days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
    for day in days:
        if day in meal_plan:
            row = [
                day,
                meal_plan[day]['Breakfast'],
                meal_plan[day]['Mid-Morning Snack'],
                meal_plan[day]['Lunch'],
                meal_plan[day]['Afternoon Snack'],
                meal_plan[day]['Dinner'],
                meal_plan[day]['Evening']
            ]
            table_data.append(row)
    
    # Create table
    # Calculate column widths - Day column narrower, others distribute remaining space
    col_widths = [0.8*inch] + [1.4*inch] * 6
    
    # Create table with Paragraph objects for text wrapping
    table_data_formatted = []
    for row_idx, row in enumerate(table_data):
        formatted_row = []
        for col_idx, cell in enumerate(row):
            if row_idx == 0:  # Header row
                style = ParagraphStyle(
                    'HeaderStyle',
                    parent=getSampleStyleSheet()['Normal'],
                    fontSize=9,
                    fontName='Helvetica-Bold',
                    textColor=colors.white,
                    alignment=1  # Center
                )
            else:
                style = ParagraphStyle(
                    'CellStyle',
                    parent=getSampleStyleSheet()['Normal'],
                    fontSize=7,
                    leading=8,
                    alignment=0  # Left
                )
            formatted_row.append(Paragraph(str(cell), style))
        table_data_formatted.append(formatted_row)
    
    table = Table(table_data_formatted, colWidths=col_widths, repeatRows=1)
    
    # Apply table style
    table.setStyle(TableStyle([
        # Header row
        ('BACKGROUND', (0, 0), (-1, 0), colors.HexColor('#2c3e50')),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.white),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('VALIGN', (0, 0), (-1, -1), 'TOP'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, 0), 9),
        
        # Alternating row colors
        ('ROWBACKGROUNDS', (0, 1), (-1, -1), [colors.white, colors.HexColor('#f8f9fa')]),
        
        # Grid
        ('GRID', (0, 0), (-1, -1), 0.5, colors.grey),
        ('LINEBELOW', (0, 0), (-1, 0), 2, colors.HexColor('#2c3e50')),
        
        # Cell padding
        ('LEFTPADDING', (0, 0), (-1, -1), 4),
        ('RIGHTPADDING', (0, 0), (-1, -1), 4),
        ('TOPPADDING', (0, 0), (-1, -1), 4),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 4),
    ]))
    
    elements.append(table)
    
    # Build PDF
    doc.build(elements)

if __name__ == '__main__':
    input_file = 'meal_plan.html'
    output_file = 'meal_plan.pdf'
    
    print(f"Parsing meal plan from {input_file}...")
    meal_plan = parse_meal_plan(input_file)
    
    print(f"Generating PDF: {output_file}...")
    create_pdf(meal_plan, output_file)
    
    print(f"✓ PDF created successfully: {output_file}")



